### ISSUE TITLE
Problem: SubsCache will not accept fast-render sub with different params.

### ISSUE THREAD

**Issue by janat08** [Blockrazor/blockrazor#1131]
Problem: SubsCache will not consider subscription ready if fast-render didn't use a subscription with the same parameters.

Solution: I imagine just making sure that all fast render subs use the same subscription parameters.

**Comment by gazhayes** [comment]
This is the core of all those other problems, right?

**Comment by janat08** [comment]
Well, the ready trigger for the infinite scroll on [/] renders like immediately incrementing parameter for subsCache :P.

**Comment by janat08** [comment]
I suppose tests could be devised for this, although $('body')[0] will always equal $('body')[0]

**Comment by gazhayes** [comment]
Yeah it's pretty hard to test for, though @anbud might have some ideas.

**Comment by gazhayes** [comment]
The simplest solution to resolve this particular issue is to crawl through the codebase and make sure all fastrender and subscache params match up. 

**The next related problem is:** this is going to keep breaking whenever someone edits params in once place and not the other.
1. use tests
2. fix subscache

@janat08 and @anbud what's the simplest of these two solutions in your opinion?

Finally, there could be one more related problem:
What about when we are using filtering on the mongo query? Does that mean we can't use subscache and fastrender there at the same time?

@janat08 Is this a problem?

**Comment by janat08** [comment]
It's not a problem since they only have to match initially, subscache only has to receive fast-render once during app load. I stubbed out subscribe calls from the router on client.

I'm unsure that it's entirely fixable as you can't necesserily tell what params are doing, but I don't like that's subsCache doesn't do everything possible to combine subscriptions once params are involved not only because that might adversely affect publications but also because there's a limit on number of subscriptions. In regards to max limit of subscriptions being easily reached I think it's a better strategy to try and detect subscriptions by page and have limit by pages since if you're subscribing for individual items having a limit is entirely unttenable for which you'd want to be able to detect when subscription is actually being used for which I already filed an issue. Anyway I'd like to improve subscache.

Back to topic at hand probably comparing subscription results from subscriptions with the same name and checking if one is subset of another would do the trick if params are numbers or same strings.

**Comment by janat08** [comment]
On another hand testing if subscache is working wouldn't be bad.

**Comment by anbud** [comment]
I believe that it would be easier to fix SubsCache than to test each SubsCache/fast-render pair, mainly because if we fix it now, we won't have to worry about tests every time we add new subscriptions.
 
Of course, if fixing SubsCache is not entirely possible, or if it's too complicated, we can always fall back to tests. We could write tests that would go through each route and check whether params of subscriptions match or we could define separate tests for each route/page.

**Comment by janat08** [comment]
I meant fast-render working rather than subscache.

**Comment by janat08** [comment]
Im guessing that would be best done by comparing fast render payloads against current subscriptions and if they're all ready.

**Comment by gazhayes** [comment]
@janat08 so the key thing here is: what's the _simplest_ way to stop people breaking this with pull requests in future?

**Comment by janat08** [comment]
tests, but I have no experience with those.

**Comment by janat08** [comment]
like what happens when you use ``fastrender.debugger.blockddp()`` which refreshes the page.

**Comment by gazhayes** [comment]
OK is the original issue resolved now? Do subs and fast render everywhere currently use the same params?

**Comment by janat08** [comment]
I saw subs that are called for specific items but don't actually pass in params to identify the item.

**Comment by janat08** [comment]
No I mean there're subs that will take id parameter to subscribe for a single item, and there're placed where these subs are used like that but don't actually pass the necessary params.

**Comment by gazhayes** [comment]
Ah ok. Do you feel like digging deeper and fixing them?

**Comment by janat08** [comment]
I think writing tests for fast-render may be easier if it can handle the debugger statement.

**Comment by janat08** [comment]
As in it's entirely possible to see all of current subscription in meteor which fast-render initiates with it's payload, and subscache subscriptions.

**Comment by gazhayes** [comment]
OK. Ping @anbud to see what's involved with that and possibly get an example test that you can use as a model to do the rest of them.

**Comment by janat08** [comment]
https://stackoverflow.com/questions/17221573/is-it-possible-to-view-all-active-subscriptions
SubsCache.cache hold subscriptions as "["subName", ...params ]"
